content/blog/2021/10/02/MonitorTheSolver.adoc

```diff
@@ -0,0 +1,215 @@
+cd optaplanner-micrometer-blog
+----
+
+The application can be started by running development mode:
+
+[source,bash]
+----
+mvn quarkus:dev
+----
+
+Once started, you can access the application at https://localhost:8080. To see available metrics for the application, visit https://localhost:8080/q/metrics.
+
+image:ExampleMonitoringApp.png[Example Application]
+
+== Starting Prometheus
+
+Prometheus has a prebuilt Docker image on Dockerhub that can be used for running Prometheus.
+We'll need to modify it configuration so it'll scrape our metrics endpoint.
+
+Create `prometheus.yml` with the following text:
+
+[source,yaml]
+----
+scrape_configs:
+- job_name: local application
+  scrape_interval: 1s
+  metrics_path: /q/metrics
+  static_configs:
+  - targets:
+    - localhost:8080
+----
+
+This configures Prometheus to scrape metrics from `localhost:8080/q/metrics` every second.
+
+To start Prometheus using the following configuration, run the command below:
+
+[source,bash]
+----
+docker run \
+    --network host \
+    --mount type=bind,source=prometheus.yml,destination=/etc/prometheus/prometheus.yml,ro=true,relabel=shared \
+     prom/prometheus
+----
+
+You can see the Prometheus UI by visiting https://localhost:9090.
+
+image::PrometheusUI.png[Prometheus UI]
+
+== Starting Grafana
+
+https://grafana.com/[Grafana] provides a much more robust UI with additional features.
+Grafana has a prebuilt image on Dockerhub you can use to run Grafana locally. Start it using the command below.
+
+[source,bash]
+----
+docker run --network host grafana/grafana
+----
+
+It may take a while to start. Once started,
+visit https://localhost:3000 to see the Grafana UI. Login with the username "admin" and the password "admin".
+
+== Connecting Grafana to Prometheus
+
+. Click on the Gear icon to go to the Configuration page.
+. Click the "Add data source" button.
+. Select "Prometheus".
+. Enter "http://localhost:9090" for the URL field.
+. (Optional) Set the Scrape Interval to be equal to the one set for Prometheus (1s).
+. Click "Save & Test".
+
+If it is set up correctly, a green textbox should appear above "Save & Test" with text "Data source is working".
+
+image::GrafanaConfigurePrometheus.png[Grafana Prometheus Configuration]
+
+== Create a dashboard to monitor metrics
+
+With all that setup, we can finally graph some metrics.
+
+. Click on the "+" icon on the left sidebar.
+. Click "Add an empty panel".
+. Beneath "A", in the text box to the right of "Metrics", enter "optaplanner_solver_solve_duration_seconds_active_count". This adds a graph for the number of active solvers. It might say "No data" if no solvers were started yet.
+. Click on the clock icon in the top right, and select "Last 5 minutes" under "Relative time range". This make the dashboard show data that occurred during the past 5 minutes.
+. Click on the arrow to the right of the Refresh icon, and click "5s". This make the dashboard refresh every 5 seconds.
```

Comment:
No such arrow on my instance of Graphana (obtained by following the steps above).

![Screen Shot 2021-09-28 at 17 04 27-fullpage](https://user-images.githubusercontent.com/673386/135114187-0770dd0d-c9a6-4030-8b7c-599c860dcef3.png)

Reply:
Odd; did the Grafana UI changed? I used latest in the docker image, but it might have used an older copy locally

Reply:
```
$ podman images grafana
REPOSITORY                 TAG         IMAGE ID      CREATED     SIZE
docker.io/grafana/grafana  latest      a38f70e6466e  7 days ago  219 MB
```

Reply:
Can confirm I was using an outdated version; look like that feature was removed from the editor :(